server/manager/src/main/java/org/apache/accumulo/manager/TabletGroupWatcher.java

```diff
@@ -257,11 +257,24 @@ public void run() {
           if (state == TabletState.ASSIGNED) {
             goal = TabletGoalState.HOSTED;
           }
+          if (Manager.log.isTraceEnabled()) {
+            Manager.log.trace(
+                "[{}] Shutting down all Tservers: {}, dependentCount: {} Extent: {}, state: {}, goal: {}",
+                store.name(), manager.serversToShutdown.equals(currentTServers.keySet()),
+                dependentWatcher == null ? "null" : dependentWatcher.assignedOrHosted(), tls.extent,
+                state, goal);
+          }
 
           // if we are shutting down all the tabletservers, we have to do it in order
           if ((goal == TabletGoalState.SUSPENDED && state == TabletState.HOSTED)
               && manager.serversToShutdown.equals(currentTServers.keySet())) {
-            if (dependentWatcher != null && dependentWatcher.assignedOrHosted() > 0) {
+            // If the stats object in the dependentWatcher is empty, then no tablets
```

Comment:
I don't quite understand the `everything is fine` is comment.  Looking at your changes I interpret the added isEmpty to rule out the case when the dependent watcher has no data about what is hosted or not.  

```
If the stats object in the dependentWatcher is empty, then it currently does not have data about what is hosted or not.  In that case host these tablets until the dependent watcher can gather some data.
```

Reply:
everything is fine == the tablet state change iterator didn't return anything == nothing needs to be done.

So, in this case nothing was returned in the previous iterator of the metadata table TGW, so `stats` was empty. In this iteration of the root table TGW dependentWatcher == the metadata table TGW and `assignedOrHosted` == 0. So, this conditional is skipped and the root tablet is unloaded.

Reply:
> I am wondering if the isEmpty() check could cause things to hang in the case when there are zero user tablets. Can not distinguish that case from not having scanned everything. Should always have a non-zero tablets at the root and metadata level, however can not assume the user level has any tablets.

If you look at how they are [constructed](https://github.com/dlmarion/accumulo/blob/3512-tserver-not-shutting-down/server/manager/src/main/java/org/apache/accumulo/manager/Manager.java#L1140), the user TGW does not have a dependent watcher.

Reply:
Root tablets will continue to be hosted while the metadata tablets are hosted/assigned or no updated information.
Metadata tables will continue to be hosted while the user tablets are hosted/assigned or no updated information.

Reply:
Updated comment in c0b9f14

Reply:
> If you look at how they are [constructed](https://github.com/dlmarion/accumulo/blob/3512-tserver-not-shutting-down/server/manager/src/main/java/org/apache/accumulo/manager/Manager.java#L1140), the user TGW does not have a dependent watcher.

I was thinking about the metadata tablets watcher which has user as a dependent.  I was wondering if the user watcher has zero tablets if its stats would always be empty meaning the metadata tablets would never unhost.

Reply:
ah, no user tables. The stats would likely be empty. I'll add a test for that and see if I can resolve that.

Reply:
Not sure if we can do this with the stats.  Wondering if we need something new, something that indicates the TGW examined all tablets in its last pass and none were hosted.

Reply:
I have a solution for the no user tables situation, will commit shortly.

Reply:
Added in 7aef709